### DISCUSSION TITLE
How to reformat istio's auth error resp without changing app's error response？

### DISCUSSION THREAD

**Discussion by elvenlegolas** [istio/istio#48986] [Q&A] [answered]
Hi,I'm working on changing istio's auth error like RBAC denied with code 403 to my own response like "you don't have permission" with code 401.The way I'm trying is using wasmplugin with a wasm written by go,but the problem is I can't distinguish istio's auth error from application's error,so the application's error will be changed either,is there anyway to identify istio's return by response header?Or is there another way to feed this demand?Thanks a lot!

**Comment by rootsongjc** [comment]
You might use response headers or status codes as indicators. 

For example, you can create an EnvoyFilter with Lua Script or Wasm to check for a 403 status code and no presence of an application-specific header (assuming your application always includes it in its responses) to identify an Istio RBAC denial. And modify the response body at the same time.

**Reply by elvenlegolas** [reply]
Firstly,thanks for replying!Sadly I can't change the application's resp head,actually I have no right to change anything from the application,that makes the situation so hard.

**Reply by elvenlegolas** [reply]
What I want to say is there is no application-specified head,and I can't add one..

**Reply by rootsongjc** [reply]
Yes. It's a bit complicated. How you distinguish them depends a bit on the logic of your application.

**Comment by zirain** [comment]
you can try with this wasmplugin https://github.com/zirain/playground/blob/master/go-playgroud/tools/wasm-extensions/mutate-body/wasmplugin.yaml

**do not use it in prod env**

**Reply by elvenlegolas** [reply]
Thanks for replaying,I saw your proxywasm code and found out you rewrote the OnHttpResponseHeader func,change resp body by identifying code403,but problem is the application would also return 403 and I don't want to reformat their response.

**Reply by zirain** [reply]
you'can also check the response body when `endOfStream==true` before envoy support this. IIRC, it's hardcode for now.

**Reply by elvenlegolas** [reply]
Sorry I kinda can't get you,may you give a more comprehensive explanation?Is there any chance to change resp code meanwhile checking resp body?Thanks a lot!

**Reply by zirain** [reply]
I updated the code, you should take a look.

**Reply by elvenlegolas** [reply]
I've read your code,but i need to modify not only resp body,but also the resp code,which is in the header.The problem is i can't change resp header in OnHttpResponsebody func,right?

**Reply by zirain** [reply]
try `proxywasm.SendHttpResponse`

**Reply by elvenlegolas** [reply]
That makes sense,I will try tomorrow,thx bro!

**Comment by elvenlegolas** [comment]
> try `proxywasm.SendHttpResponse`

It doesn't work…I saw the original code of that func,it cannot be used after resp header is sent,which means can't be invoked in OnHttpResponseBody.

**Reply by zirain** [reply]
you can try to overwide the headers if possible, otherwise I have no idea for now.

**Reply by zirain** [reply]
@elvenlegolas from following demo, it seems fine to call SendHttpResponse on OnHttpRequestBody

https://github.com/tetratelabs/proxy-wasm-go-sdk/blob/db22d5a6f655e97f3f02790df4ee318250a00330/examples/http_body_chunk/main.go#L86

**Reply by elvenlegolas** [reply]
Yeah that's ok,but notice that it's **OnHttpRequestBody**,yet I need to invoke it during **OnHttpResponseBody**.,I'm trying to use lua script instead of wasm rn,I will catch you up if there is any result.

**Reply by zirain** [reply]
oh, I miss that part.

**Reply by elvenlegolas** [reply, marked as answer]
Hey, I've finally found a solution that involves using an Envoy filter with a Lua script. Here's the code:

```yaml
defaultSourceCode:
  inlineString: |
    function envoy_on_response(response_handle)
      -- Check the response body and modify Istio's authentication error response
      local body = response_handle:body()
      local errMap = {["Jwt is expired"] = 0, ["RBAC: access denied"] = 0}
      local bodyStr = body:getBytes()
      if errMap[bodyStr] == 0 then
        local headers = response_handle:headers()
        headers:replace(":status", "401")
        body:setBytes("iAuthenticator idmAuthResult false")
      end
    end
```
